src/client/coll/mod.rs

```diff
@@ -0,0 +1,309 @@
+        Ok(InsertManyResult::new(res))
+    }
+
+    // Internal deletion helper function.
+    fn delete(&self, filter: bson::Document, limit: i64, write_concern: Option<WriteConcern>) -> Result<DeleteResult, String> {
+        let wc = match write_concern {
+            Some(wc) => wc,
+            None => WriteConcern::new(),
+        };
+
+        let mut deletes = bson::Document::new();
+        deletes.insert("q".to_owned(), Bson::Document(filter));
+        deletes.insert("limit".to_owned(), Bson::I64(limit));
+
+        let mut cmd = bson::Document::new();
+        cmd.insert("delete".to_owned(), Bson::String(self.name()));
+        cmd.insert("deletes".to_owned(), Bson::Array(vec!(Bson::Document(deletes))));
+        cmd.insert("writeConcern".to_owned(), Bson::Document(wc.to_bson()));
+
+        let res = try!(self.db.command(cmd));
+        match res {
+            Some(doc) => Ok(DeleteResult::new(doc)),
+            None => Err("Delete reply not received from server.".to_owned()),
+        }
+    }
+
+    /// Deletes a single document.
+    pub fn delete_one(&self, filter: bson::Document, write_concern: Option<WriteConcern>) -> Result<DeleteResult, String> {
+        self.delete(filter, 1, write_concern)
+    }
+
+    /// Deletes multiple documents.
+    pub fn delete_many(&self, filter: bson::Document, write_concern: Option<WriteConcern>) -> Result<DeleteResult, String> {
+        self.delete(filter, 0, write_concern)
+    }
+
+    // Internal update helper function.
+    fn update(&self, filter: bson::Document, update: bson::Document, upsert: bool, multi: bool,
+              write_concern: Option<WriteConcern>) -> Result<UpdateResult, String> {
+
+        let wc = match write_concern {
+            Some(wc) => wc,
+            None => WriteConcern::new(),
+        };
+
+        let mut updates = bson::Document::new();
+        updates.insert("q".to_owned(), Bson::Document(filter));
+        updates.insert("u".to_owned(), Bson::Document(update));
+        updates.insert("upsert".to_owned(), Bson::Boolean(upsert));
+        updates.insert("multi".to_owned(), Bson::Boolean(multi));
+
+        let mut cmd = bson::Document::new();
+        cmd.insert("update".to_owned(), Bson::String(self.name()));
+        cmd.insert("updates".to_owned(), Bson::Array(vec!(Bson::Document(updates))));
+        cmd.insert("writeConcern".to_owned(), Bson::Document(wc.to_bson()));
+
+        let res = try!(self.db.command(cmd));
+        match res {
+            Some(doc) => Ok(UpdateResult::new(doc)),
+            None => Err("delete_many reply not received from server.".to_owned()),
+        }
+    }
+
+    /// Replaces a single document.
+    pub fn replace_one(&self, filter: bson::Document, replacement: bson::Document, upsert: bool,
+                       write_concern: Option<WriteConcern>) -> Result<UpdateResult, String> {
+
+        for key in replacement.keys() {
+            if key.starts_with("$") {
```

Comment:
Seems like a fairly harmless check. This is also the only thing that differentiates replace and update, so it may prevent mix-ups if someone uses `update` and accidentally provides a replace document.